waspc/src/Wasp/GeneratorMonad.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Wasp.GeneratorMonad
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    GeneratorState (..),
+    catchGeneratorError,
+    logGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the fatal error that halted generation.
+-- There may optionally be other errors or warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError
+  = GenericGeneratorError String
+  | PackageJsonGenerationError String
+  | PrismaGenerationError String
```

Comment:
Yeah good point, this may be opening a can of worms. I can nix them all now and only have Generic for both error and warning until we see a need for a caller to catch/match on it. 👍🏻

Reply:
Removed all specific errors for now 👍🏻